shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java

```diff
@@ -828,10 +827,9 @@ public static WireCommand readFrom(ByteBufInputStream in, int length) throws IOE
             if (dataLength > length) {
                 throw new BufferOverflowException();
             }
-            byte[] data = new byte[dataLength];
-            in.readFully(data);
-            long requestId =  in.available() >= Long.BYTES ? in.readLong() : -1L;
-            return new SegmentRead(segment, offset, atTail, endOfSegment, ByteBuffer.wrap(data), requestId);
+            ByteBuf data = in.readFully(dataLength).retain();
```

Comment:
This is adding a call to `retain()`. Where is the corresponding `release()` call?

Reply:
Good point. Due to the fact that this class is instantiated in 2 ways (with an external ByteBuf via the constructor, and by deserializing), I only need to retain for the serialization, so the solution is a bit less elegant than I wished. I added some unit tests to verify it works as I intend to.

The corresponding `release` has been added to `SegmentInputStreamImpl.handleRequest`

Reply:
I am not 100% sure this is correct. But I do not know of any way to prevent memory leaks without making a copy in the deserialization. I am concerned about cancelled requests or other unexpected situations.

Reply:
SegmentInputStreamImpl:89 is missing the logic to call release. In case setoffset is called before reading the data is complete.